drivers/dai/intel/ssp/ssp.c

```diff
@@ -1913,52 +1915,35 @@ static int dai_ssp_remove(struct dai_intel_ssp *dp)
 	return 0;
 }
 
-static int dai_ssp_probe_wrapper(const struct device *dev)
+static int ssp_pm_action(const struct device *dev, enum pm_device_action action)
 {
 	struct dai_intel_ssp *dp = (struct dai_intel_ssp *)dev->data;
-	k_spinlock_key_t key;
-	int ret = 0;
-
-	key = k_spin_lock(&dp->lock);
-
-	if (dp->sref == 0) {
-		ret = dai_ssp_probe(dp);
-	}
-
-	if (!ret) {
-		dp->sref++;
+	switch (action) {
+	case PM_DEVICE_ACTION_SUSPEND:
+		dai_ssp_remove(dp);
+		break;
+	case PM_DEVICE_ACTION_RESUME:
+		dai_ssp_probe(dp);
+		break;
+	default:
+		return -ENOTSUP;
 	}
 
```

Comment:
Shouldn't it handle `PM_DEVICE_ACTION_TURN_ON|OFF` ? I mean, wouldn't it be under a power domain ? `TURN_ON` and `TURN_OFF` are actions sent when the power domain the driver belongs is being resumed or suspended.

Reply:
I don't think so, there's nothing else needed to be done except for what is done while suspending. We should only use zephyrs pd logic the other way round, so when no devices are in use - turn off the domain. The power there is controlled through ipc messages. We could try to force the device to suspend if the pd is being turned off while the device is running, but this scenario would break stuff on the higher level anyway.

Reply:
gotcha, I still think it is worth to handle those `TURN_ON` and `TURN_OFF` actions here, even if it is just to add an assert saying that the action is not expected. This makes easier to catch bugs. And technically speaking, this is under a domain that when suspended / resume will send this notifications to here and if you return an error code the subsystem may interpret is an error.